### DISCUSSION TITLE
await waitForTransactionReceipt() doesn't wait enough long and it executes next commands before the metamask displays completed confirmation

### DISCUSSION THREAD

**Discussion by InfiniteAengus** [wevm/wagmi#3789] [Q&A]
Before calling the function of the smart contract, I have to approve the token to the contract.

```
const tx = await writeContract(config, {
			abi: erc20ABI,
			address: tokenAddress,
			functionName: 'approve',
			args: [contractAddress, swapValue],
		})
await waitForTransactionReceipt(config, { hash: tx })

await writeContract(config, {
				abi: contractABI,
				address: contractAddress,
				functionName: 'abc',
				args:  [tokenAddress, aAddress, amount]
			})
```

But the problem is calling the 'abc' function is executed before the token is approved.
I click the approve in the metamask, after a few seconds, the second pop up is appeared for 'abc' function.
But approve is not completed yet, and the metamask shows there's one pending transaction and since the token is not approved yet, the function call reverted.

What's wrong with my code or how can I fix this problem?

**Comment by tmm** [comment]
What RPC are you using?
